javatests/arcs/sdk/HandleHolderBaseTest.kt

```diff
@@ -0,0 +1,213 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+package arcs.sdk
+
+import com.google.common.truth.Truth.assertThat
+import com.nhaarman.mockitokotlin2.any
+import com.nhaarman.mockitokotlin2.doReturn
+import com.nhaarman.mockitokotlin2.mock
+import com.nhaarman.mockitokotlin2.times
+import com.nhaarman.mockitokotlin2.verify
+import kotlin.test.assertFailsWith
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.test.runBlockingTest
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.junit.runners.JUnit4
+
+@OptIn(ExperimentalCoroutinesApi::class)
+@RunWith(JUnit4::class)
+class HandleHolderBaseTest {
+
+  @Test
+  fun handles_accessMissingHandle_throwsNoElemException() {
+    val handleHolder = DummyHandleHolder()
+
+    assertFailsWith<NoSuchElementException> {
+      handleHolder.readHandle
+    }.also {
+      assertThat(it.message).isEqualTo(
+        "Handle readHandle has not been initialized in TestParticle yet."
+      )
+    }
+  }
+
+  @Test
+  fun handles_nameNotInEntitySpec_throwsNoElemException() {
+    val handleHolder = DummyHandleHolder()
+
+    assertFailsWith<NoSuchElementException> {
+      handleHolder.handleButNotEntities
+    }.also {
+      assertThat(it.message).isEqualTo(
+        "Particle TestParticle does not have a handle with name handleButNotEntities."
+      )
+    }
+  }
+
+  @Test
+  fun dispatcher_noHandles_asserts() {
+    val emptyHandleHolder = DummyHandleHolder()
+
+    assertFailsWith<IllegalStateException> {
+      emptyHandleHolder.dispatcher
+    }.also {
+      assertThat(it.message).isEqualTo(
+        "No dispatcher available for a HandleHolder with no handles."
+      )
+    }
+  }
+
+  @Test
+  fun dispatcher_someHandles_getsFirstHandlesDispatcher() {
+    val handleHolder = DummyHandleHolder()
+    val firstHandle = mock<Handle> {
+      on { dispatcher } doReturn DUMMY_DISPATCHER_1
+    }
+    val secondHandle = mock<Handle> {
+      on { dispatcher } doReturn DUMMY_DISPATCHER_2
```

Comment:
This one is interesting... I think technically it should be an error to try and configure a HandleHolder this way. All handles should have the same dispatcher. I wonder if we should consider adding something that verifies that as handles are set... (not in this PR of course)

Reply:
Ah, I didn't realize that this is how dispatchers are actually used. I can change to test the interface and not the implementation.

Reply:
Well, your test is "technically correct" right now. I think the issue is really that we should fix the class to ensure that invariant...